wrappers/golang/core/poseidon.go

```diff
@@ -0,0 +1,98 @@
+package core
+
+import (
+	"fmt"
+	"unsafe"
+
+	cr "github.com/ingonyama-zk/icicle/v2/wrappers/golang/cuda_runtime"
+)
+
+type PoseidonConfig struct {
+	/// Details related to the device such as its id and stream id. See [DeviceContext](@ref device_context::DeviceContext).
+	Ctx                cr.DeviceContext
+	areInputsOnDevice  bool
+	areOutputsOnDevice bool
+	///If true, input is considered to be a states vector, holding the preimages in aligned or not aligned format.
+	///Memory under the input pointer will be used for states. If false, fresh states memory will be allocated and input will be copied into it */
+	InputIsAState bool
+	/// If true - input should be already aligned for poseidon permutation.
+	///* Aligned format: [0, A, B, 0, C, D, ...] (as you might get by using loop_state)
+	///* not aligned format: [A, B, 0, C, D, 0, ...] (as you might get from cudaMemcpy2D) */
+	Aligned bool
+	///If true, hash results will also be copied in the input pointer in aligned format
+	LoopState bool
+	///Whether to run the Poseidon asynchronously. If set to `true`, the poseidon_hash function will be
+	///non-blocking and you'd need to synchronize it explicitly by running `cudaStreamSynchronize` or `cudaDeviceSynchronize`.
+	///If set to false, the poseidon_hash function will block the current CPU thread. */
+	IsAsync bool
+}
+
+type PoseidonConstants[T any] struct {
+	Arity           int32
+	PartialRounds   int32
+	FullRoundsHalf  int32
+	RoundConstants  unsafe.Pointer
+	MdsMatrix       unsafe.Pointer
+	NonSparseMatrix unsafe.Pointer
+	SparseMatrices  unsafe.Pointer
+	DomainTag       T
+}
+
+func GetDefaultPoseidonConfig() PoseidonConfig {
+	ctx, _ := cr.GetDefaultDeviceContext()
+	return PoseidonConfig{
+		ctx,   // Ctx
+		false, // areInputsOnDevice
+		false, // areOutputsOnDevice
+		false, // inputIsAState
+		false, // aligned
+		false, // loopState
+		false, // IsAsync
+	}
+}
+
+func PoseidonCheck[T any](input, output HostOrDeviceSlice, cfg *PoseidonConfig, constants *PoseidonConstants[T], arity, numberOfStates int) (unsafe.Pointer, unsafe.Pointer, unsafe.Pointer) {
+	inputLen, outputLen := input.Len(), output.Len()
+
+	if arity != int(constants.Arity) {
+		errorString := fmt.Sprintf(
+			"arity: %d does not match Poseidon Constants: %d",
+			arity,
+			outputLen,
+		)
+		panic(errorString)
+	}
+
+	if inputLen != arity*numberOfStates {
```

Comment:
Does the expected input length depend on whether `cfg.InputIsAState` is true? See https://github.com/ingonyama-zk/icicle/blob/main/wrappers/rust/icicle-core/src/poseidon/mod.rs#L160-L166
cc @ChickenLover

Reply:
Should we allow an input length that is larger than `arity*numStates`

Reply:
> Does the expected input length depend on whether `cfg.InputIsAState` is true? See https://github.com/ingonyama-zk/icicle/blob/main/wrappers/rust/icicle-core/src/poseidon/mod.rs#L160-L166 cc @ChickenLover

I see, I'll fix it

Reply:
> Should we allow an input length that is larger than `arity*numStates`

well the end of the slice won't be taken into account right ? If users don't use it well you might end up with a hash where you can easily get a collision, checking that it is equal seems safer even if not strictly necessary